### PR TITLE
Fix docker image build for `Latest PyTorch + TensorFlow [dev]`

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

It's unknown why the image/layers could not be exported and pushed. The best guess is that it is way too large for which I don't know the limitation.

This PR combines several steps in one single `RUN ...` command in the dockerfile so the size is smaller.

Successful build: https://github.com/huggingface/transformers/actions/runs/8365229494/job/22902558338

### CODE REVIEW

File: docker/transformers-all-latest-gpu/Dockerfile [hunk 1]
```diff
@@ -9,7 +9,7 @@ SHELL ["sh", "-lc"]
 # The following `ARG` are mainly used to specify the versions explicitly & directly in this docker file, and not meant
 # to be used as arguments for docker build (so far).
 
-ARG PYTORCH='2.2.0'
+ARG PYTORCH='2.2.1'
```
Comment (ydshieh) [inline docker/transformers-all-latest-gpu/Dockerfile:12]: let's use the latest one

File: docker/transformers-all-latest-gpu/Dockerfile [hunk 2]
```diff
@@ -9,7 +9,7 @@ SHELL ["sh", "-lc"]
...
 # (not always a valid torch version)
 ARG INTEL_TORCH_EXT='2.2.0'
```
Comment (ydshieh) [inline docker/transformers-all-latest-gpu/Dockerfile:14]: no `2.2.1` available for `INTEL_TORCH_EXT` yet. But I guess it's fine

File: docker/transformers-all-latest-gpu/Dockerfile [hunk 3]
```diff
@@ -23,21 +23,7 @@ RUN python3 -m pip install --no-cache-dir --upgrade pip
...
-# Currently, let's just use their latest releases (when `torch` is installed with a release version)
-# TODO: We might need to specify proper versions that work with a specific torch version (especially for past CI).
-RUN [ "$PYTORCH" != "pre" ] && python3 -m pip install --no-cache-dir -U $VERSION torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/$CUDA || python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/$CUDA
+RUN python3 -m pip install --no-cache-dir -U tensorflow==2.13 protobuf==3.20.3 tensorflow_text tensorflow_probability && python3 -m pip install --no-cache-dir -e ./transformers[dev,onnxruntime] && [ ${#PYTORCH} -gt 0 -a "$PYTORCH" != "pre" ] && VERSION='torch=='$PYTORCH'.*' ||  VERSION='torch'; echo "export VERSION='$VERSION'" >> ~/.profile && echo torch=$VERSION && [ "$PYTORCH" != "pre" ] && python3 -m pip install --no-cache-dir -U $VERSION torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/$CUDA || python3 -m pip install --no-cache-dir -U --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/$CUDA
```
Comment (ydshieh) [inline docker/transformers-all-latest-gpu/Dockerfile:26]: put everything in a single `RUN ...` to reduce the layers and size.

**Comment by ydshieh** [comment]
Hm, the last run on `main` works now despite this PR hasn't been merged.

File: docker/transformers-all-latest-gpu/Dockerfile (hunk 3 above)
Comment (amyeroberts) [reply docker/transformers-all-latest-gpu/Dockerfile:26]: This is such a hack 😅 

I think we still want some of the comments to provide context

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

File: docker/transformers-all-latest-gpu/Dockerfile (hunk 3 above)
Comment (ydshieh) [reply docker/transformers-all-latest-gpu/Dockerfile:29]: Added back (part of) the previous comments

**Pull request merged**